docs(oidc): supporting dpop

StarlightIbuki opened the pull request:
### Description

KAG-4377

New feature support.

https://github.com/Kong/kong-ee/pull/8482

### Testing instructions

Preview link: https://deploy-preview-7323--kongdocs.netlify.app/hub/kong-inc/openid-connect/unreleased/how-to/demonstrating-proof-of-possession/

### Checklist 

- [x] Review label added <!-- (see below) -->
- [x] [Conditional version tags](https://docs.konghq.com/contributing/conditional-rendering/#conditionally-render-content-by-version) added, if applicable.

<!-- For example, if this change is for an upcoming 3.6 release, enclose your content in `{% if_version gte:3.6.x %} <content> {% endif_version %}` tags (or `if_plugin_version` tags for plugins). 

Use any of the following keys:
* `gte:<version>` - greater than or equal to a specific version
* `lte:<version>` - less than or equal to a specific version
* `eq:<version>` - exactly equal to a specific version

You can do the same for older versions. -->

<!-- !!! Only Kong employees can add labels due to a GitHub limitation. If you're an OSS contributor, thank you! The maintainers will label this PR for you !!! -->

<!-- When raising a pull request, indicate what type of review you need with one of the following labels:

    review:copyedit: Request for writer review.
    review:general: Review for general accuracy and presentation. Does the doc work? Does it output correctly?
    review:tech: Request for technical review for a docs platform change.
    review:sme: Request for review from an SME (engineer, PM, etc).

At least one of these labels must be applied to a PR or the build will fail.
-->

lena-larionova commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating_proof_of_possession.md:
````diff
@@ -0,0 +1,103 @@
+1. To enable Demonstrating Proof-of-Possession, use the [`proof_of_possession_dpop`](/hub/kong-inc/openid-connect/configuration/#config-proof_of_possession_dpop) configuration option:
+
+    ```bash
+    http -f put :8001/plugins/5f35b796-ced6-4c00-9b2a-90eef745f4f9 \
````
Can you convert this to cURL? We try to avoid using httpie since it's a separate tool that someone would have to install, whereas we can assume they have curl.

lena-larionova commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating_proof_of_possession.md:
````diff
@@ -0,0 +1,103 @@
+
+1. To enable Demonstrating Proof-of-Possession, use the [`proof_of_possession_dpop`](/hub/kong-inc/openid-connect/configuration/#config-proof_of_possession_dpop) configuration option:
+
+    ```bash
+    http -f put :8001/plugins/5f35b796-ced6-4c00-9b2a-90eef745f4f9 \
+    name=openid-connect                                          \
+    service.name=openid-connect                                  \
+    config.issuer=https://keycloak.test:8440/auth/realms/master  \
+    config.client_id=cert-bound                                  \
+    config.client_secret=cf4c655a-0622-4ce6-a0de-d3353ef0b714    \
+    config.auth_methods=bearer                                   \
+    config.proof_of_possession_dpop=strict
+    ```
````
Let's use this formatting:

https://github.com/Kong/docs.konghq.com/blob/main/app/_hub/kong-inc/openid-connect/how-to/client-authentication/_mtls.md?plain=1#L77-L96

That will generate a full example with all selected tools/formats: 
https://docs.konghq.com/hub/kong-inc/openid-connect/how-to/client-authentication/mtls/#set-up-mtls-for-the-oidc-plugin

lena-larionova commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating_proof_of_possession.md:
```diff
@@ -0,0 +1,103 @@
+## Overview
+
+Demonstrating Proof-of-Possession(DPoP) is an alternative technique to the Mutual TLS certificate-bound access tokens. Instead of binding the token with the client certificate used in the mTLS, it binds the token with a JWK provided by the client.
+
```
Add diagram here. Proposed draft:

```
<!--vale off-->
{% mermaid %}
sequenceDiagram
    autonumber
    participant client as Client <br>(e.g. mobile app)
    participant kong as API Gateway <br>(Kong)
    participant httpbin as Upstream <br>(backend service,<br> e.g. httpbin)
    activate client
    client->>client: generate key pair
    activate kong
    client->>kong: POST /oauth2/token<br>DPoP:$PROOF
    kong-->>client: DPoP bound access token ($AT)
    deactivate kong
    activate httpbin
    client->>httpbin: GET https://example.com/resource<br>Authorization: DPoP $AT<br>DPoP: $PROOF
    deactivate client
    httpbin->>httpbin: validate $AT and $PROOF
    activate client
    httpbin-->>client: response
    deactivate client
    deactivate httpbin
{% endmermaid %}
<!--vale on-->
```

Will render as something like this:

![Screenshot 2024-05-02 at 10 11 05 PM](https://github.com/Kong/docs.konghq.com/assets/54370747/746c671f-affc-45b5-b7d7-417933e7e3b7)

lena-larionova commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating_proof_of_possession.md:
```diff
@@ -0,0 +1,103 @@
+
+## Overview
+
+Demonstrating Proof-of-Possession(DPoP) is an alternative technique to the Mutual TLS certificate-bound access tokens. Instead of binding the token with the client certificate used in the mTLS, it binds the token with a JWK provided by the client.
```
```suggestion
Demonstrating Proof-of-Possession (DPoP) is an alternative technique to the [Mutual TLS certificate-bound access tokens](/hub/kong-inc/openid-connect/how-to/client-authentication/mtls/). Instead of binding the token with the client certificate used in the mTLS, it binds the token with a JWK provided by the client.
```

lena-larionova commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating_proof_of_possession.md:
```diff
@@ -0,0 +1,103 @@
+
+Demonstrating Proof-of-Possession(DPoP) is an alternative technique to the Mutual TLS certificate-bound access tokens. Instead of binding the token with the client certificate used in the mTLS, it binds the token with a JWK provided by the client.
+
+It's possible to use Demonstrating Proof-of-Possession without mTLS, or even with a plain HTTP. But HTTPS is recommended for better protection.
```
```suggestion
It's possible to use Demonstrating Proof-of-Possession without mTLS, or even with plain HTTP. HTTPS is recommended for better protection.
```

lena-larionova commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating_proof_of_possession.md:
```diff
@@ -0,0 +1,103 @@
+
+It's possible to use Demonstrating Proof-of-Possession without mTLS, or even with a plain HTTP. But HTTPS is recommended for better protection.
+
+Note that by enabling the verification of the DPoP proof, Kong removes the `DPoP` header and changes the token type from "dpop" to "bearer", which essentially downgrades the request to use a conventional bearer token.
```
```suggestion
Note that by enabling the verification of the DPoP proof, Kong removes the `DPoP` header and changes the token type from `dpop` to `bearer`, which essentially downgrades the request to use a conventional bearer token.
```

lena-larionova commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating_proof_of_possession.md:
```diff
@@ -0,0 +1,103 @@
+It's possible to use Demonstrating Proof-of-Possession without mTLS, or even with a plain HTTP. But HTTPS is recommended for better protection.
+
+Note that by enabling the verification of the DPoP proof, Kong removes the `DPoP` header and changes the token type from "dpop" to "bearer", which essentially downgrades the request to use a conventional bearer token.
+This allows an OAuth2 upstream without DPoP support to work with the DPoP token without loosing the protection of key-binding mechanism.
```
```suggestion
This allows an OAuth2 upstream without DPoP support to work with the DPoP token without losing the protection of the key-binding mechanism.
```

lena-larionova commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating_proof_of_possession.md:
```diff
@@ -0,0 +1,103 @@
+Enabling the [`proof_of_possession_dpop`](/hub/kong-inc/openid-connect/configuration/#config-proof_of_possession_dpop) configuration option in the plugin helps to ensure that the supplied access token
+belongs to the client by verifying its binding with the client certificate provided in the request.
+
+The DPoP are supported by the following auth methods:
```
```suggestion
DPoP is supported by the following auth methods:
```

lena-larionova commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating_proof_of_possession.md:
````diff
@@ -0,0 +1,103 @@
+    }
+    ```
+
+1. Obtain the token from the IdP. A DPoP supporting client is required. For the demonstration purpose below is an example using [`oauth2c`](https://github.com/cloudentity/oauth2c.git):
````
```suggestion
1. Obtain the token from the IdP. A DPoP supporting client is required. For this demo, here is an example using [`oauth2c`](https://github.com/cloudentity/oauth2c.git):
```

lena-larionova reviewed (commented):
@StarlightIbuki I converted the example into the same format used in our other OIDC guides, which generates output for multiple tools & formats. 

Also added a diagram based on info from @veenaraja, adapted to Kong - please take a look and let me know if it's accurate for our use case.

StarlightIbuki commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating_proof_of_possession.md:
````diff
@@ -0,0 +1,148 @@
+```
+
+Access the service using the same client certificate and key used to obtain the token.
+The client should generate proper proof for possession of the key and send it via the `DPoP` header, which will be verified by Kong.
````
`oauth2c` cannot be used to initiate an RP request thus we cannot use it in this step. I cannot find a proper 3rd party tool to do the demo here.

StarlightIbuki commented:
Should we also add a reference from `Mutual TLS certificate-bound access tokens` to this? @lena-larionova

StarlightIbuki commented:
> @StarlightIbuki I converted the example into the same format used in our other OIDC guides, which generates output for multiple tools & formats.
> 
> Also added a diagram based on info from @veenaraja, adapted to Kong - please take a look and let me know if it's accurate for our use case.

2 things are different in the use of the current design of this feature:
1. The token is not necessarily issued (proxied) by Kong;
2. We're downgrading the DPoP token thus the proof and `DPoP` token type are invisible from the upstream.

I will fix the diagram.

Guaris commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating-proof-of-possession.md:
```diff
@@ -0,0 +1,155 @@
+
+It's a prerequisite that the auth server is configured to enable the DPoP for you to enable it with Kong.
+
+If you are configuring this in Keycloak, see the [Keycloak configuration](#prerequisites) section in the prerequisites.
+For alternative auth servers, consult their documentation to configure this functionality.
```
```suggestion
If you are setting up Demonstrating Proof of Possession in Keycloak, refer to the   [Keycloak configuration](#prerequisites) section in the prerequisites.
For other authentication servers, please consult their specific documentation to configure this functionality.
```

Guaris commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating-proof-of-possession.md:
```diff
@@ -0,0 +1,155 @@
+If you are configuring this in Keycloak, see the [Keycloak configuration](#prerequisites) section in the prerequisites.
+For alternative auth servers, consult their documentation to configure this functionality.
+
+Some of the instructions in the other how-to guides for OpenID Connect support validation of access tokens using mTLS proof of possession.
+Enabling the [`proof_of_possession_dpop`](/hub/kong-inc/openid-connect/configuration/#config-proof_of_possession_dpop) configuration option in the plugin helps to ensure that the supplied access token
+belongs to the client by verifying its binding with the client certificate provided in the request.
```
```suggestion
Many how-to guides for OpenID Connect include instructions for validating access tokens using mTLS proof of possession.
Enabling the [`proof_of_possession_dpop`](/hub/kong-inc/openid-connect/configuration/#config-proof_of_possession_dpop) configuration option in the plugin to ensures that the access token
is bound to the client by verifying its association with the client certificate provided in the request.
```

Guaris commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating-proof-of-possession.md:
```diff
@@ -0,0 +1,155 @@
+Enabling the [`proof_of_possession_dpop`](/hub/kong-inc/openid-connect/configuration/#config-proof_of_possession_dpop) configuration option in the plugin helps to ensure that the supplied access token
+belongs to the client by verifying its binding with the client certificate provided in the request.
+
+DPoP is supported by the following auth methods:
```
```suggestion
DPoP is compatible with the following authentication methods:
```

Guaris commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating-proof-of-possession.md:
```diff
@@ -0,0 +1,155 @@
+
+   Session Authentication is only compatible with DPoP when used along with one of the other supported authentication methods:
+
+    * If multiple `openid-connect` plugins are configured with the `session` auth method, we strongly recommend configuring different values of [`session_secret`](/hub/kong-inc/openid-connect/configuration/#config-session_secret) across plugin instances for additional security. This avoids sessions being shared across plugins and possibly bypassing the proof of possession validation.
```
```suggestion
    * If multiple `openid-connect` plugins are configured with the `session` authentication method, we strongly recommend configuring different values of [`session_secret`](/hub/kong-inc/openid-connect/configuration/#config-session_secret) across plugin instances for additional security. This avoids sessions being shared across plugins and possibly bypassing the proof of possession validation.
```

Guaris commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating-proof-of-possession.md:
```diff
@@ -0,0 +1,155 @@
+{% endplugin_example %}
+<!-- vale on -->
+
+If this is configured correctly, it returns a `200` response and something like the following:
```
```suggestion
If configured correctly, it returns a `200` response like this:
```

Guaris commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating-proof-of-possession.md:
````diff
@@ -0,0 +1,155 @@
+--signing-key https://path/to/the_jwks.json
+```
+
+If this is configured correctly, it returns a `200` response and something like the following:
````
```suggestion
If this is configured correctly, it returns a `200` response like this:
```

Guaris commented on app/_includes/md/plugins-hub/oidc-prereqs.md:
```diff
@@ -22,15 +22,19 @@ difficulties during this phase, refer to the [Keycloak documentation](https://ww
    From the **Credentials** tab, select the **X509 Certificate** Client Authenticator and fill the Subject DN field so that it matches the Kong client certificate's, e.g.: `CN=JohnDoe, OU=IT`.
 {% endif_plugin_version %}
+{% if_plugin_version gte:3.7.x %}
+1. (Optional, to test Demonstrating Proof-of-Possession Client Authentication) Create another confidential client `client-dpop-auth` with settings similar to the `service` client created above. 
```
```suggestion
1. **Optional**: To test Demonstrating Proof-of-Possession Client Authentication, create another confidential client `client-dpop-auth` with settings similar to the `service` client created above. 
```

StarlightIbuki replied:
We have a line in this format before the change, so we need to keep consistent with it, or we need to change them altogether.

StarlightIbuki commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating-proof-of-possession.md:
```diff
@@ -0,0 +1,155 @@
+
+Demonstrating Proof-of-Possession (DPoP) is an alternative technique to the 
+[Mutual TLS certificate-bound access tokens](/hub/kong-inc/openid-connect/how-to/client-authentication/mtls/). 
+Instead of binding the token with the client certificate used in the mTLS, it binds the token with a JWK provided by the client.
```
The rephrased sentence is of a slightly different meaning. "used in the mTLS" describes the cert, and DPoP cannot be compared with mTLS.

StarlightIbuki commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating-proof-of-possession.md:
```diff
@@ -0,0 +1,155 @@
+{% endmermaid %}
+<!--vale on-->
+
+It's possible to use Demonstrating Proof-of-Possession option without mTLS, or even with plain HTTP. HTTPS is recommended for better protection.
+
```
"Demonstrating Proof-of-Possession" is the standard written form.

StarlightIbuki commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating-proof-of-possession.md:
```diff
@@ -0,0 +1,155 @@
+
+It's possible to use Demonstrating Proof-of-Possession option without mTLS, or even with plain HTTP. HTTPS is recommended for better protection.
+
+Note that by enabling the verification of the DPoP proof, Kong removes the `DPoP` header and changes the token type from `dpop` to `bearer`,
```
The standard capitalization of the header is "DPoP" although we accept different capitalization.

Guaris replied:
Fixed @StarlightIbuki

StarlightIbuki commented:
@lena-larionova Hi. Could you review it again and see if the PR is ok to merge?

lena-larionova commented on app/_hub/kong-inc/openid-connect/how-to/_demonstrating-proof-of-possession.md:
```diff
@@ -0,0 +1,155 @@
+    participant client as Client <br>(e.g. mobile app)
+    participant kong as API Gateway <br>(Kong)
+    participant upstream as Upstream <br>(backend service,<br> e.g. httpbin)
+    participant idp as Authenticate Server <br>(e.g. Keycloak)
```
```suggestion
    participant idp as Authentication Server <br>(e.g. Keycloak)
```

lena-larionova reviewed (approved):
One tiny typo fix, otherwise lgtm!
